libp2p/network/swarm.py

```diff
@@ -87,17 +88,53 @@ def set_stream_handler(self, stream_handler: StreamHandlerFn) -> None:
         try:
             # Get peer info from peer store
             addrs = self.peerstore.addrs(peer_id)
-        except PeerStoreError:
-            raise SwarmException(f"No known addresses to peer {peer_id}")
+        except PeerStoreError as error:
+            raise SwarmException(f"No known addresses to peer {peer_id}") from error
 
-        if not addrs:
-            raise SwarmException(f"No known addresses to peer {peer_id}")
+        exceptions: List[SwarmException] = []
+
+        # Try all known addresses
+        for multiaddr in addrs:
+            try:
+                return await self.dial_addr(multiaddr, peer_id)
+            except SwarmException as e:
+                exceptions.append(e)
+                logger.debug(
+                    "encountered swarm exception when trying to connect to %s, "
+                    "trying next address...",
+                    multiaddr,
+                    exc_info=e,
+                )
+
+        # Tried all addresses, raising exception.
+        if len(exceptions) > 0:
+            raise SwarmException(
+                "unable to connect to %s, no addresses established a successful connection "
+                "(with exceptions)",
+                peer_id,
+            ) from MultiError(exceptions)
+        else:
+            raise SwarmException(
+                "unable to connect to %s, no addresses established a successful connection "
```

Comment:
ditto

Reply:
Besides, we should never enter here, right? If `len(exceptions) <= 0`, it means we got no exception when performing `dial_addr`. The only possible situation is `len(addrs) == 0`, which can be checked by `if not addrs:`.

Reply:
I'll change the `addr` variable to have a type hinting of it being an `Multiaddr` iterative/generator. I'll make `exceptions` be the check as to if an iteration has been run at all, and if not, change the wording on the second one as to point directly to that. (With a note in a comment that tells why exactly the program only gets there when it did not encounter any addresses in the generator)

Reply:
> I'll change the addr variable to have a type hinting of it being an Multiaddr iterative/generator.

Could you elaborate more on the reason that `addrs` should be an iterator/generator? IMO the deleted `if not addrs:` above can already sufficiently replace the `else` branch here and it's more explicit to me.

Reply:
Alright, then I'll do that.